packages/grid/src/vaadin-grid.js

```diff
@@ -667,12 +638,72 @@ class Grid extends ElementMixin(
     }
 
     this.__intrinsicWidthCache = new Map();
+    // Cache the viewport rows to avoid unnecessary reflows while measuring the column widths
+    this.__viewportRowsCache = this.__getViewportRows();
+
+    // Pre-cache the intrinsic width of each column
+    this.__calculateAndCacheIntrinsicWidths(cols);
 
     cols.forEach((col) => {
       col.width = `${this.__getDistributedWidth(col)}px`;
     });
   }
 
+  /**
```

Comment:
There is so much logic related to measuring column auto-width in `vaadin-grid.js` now that it should be extracted to a separate mixin. The change should be done in a separate PR, however.